proposals/NNNN-spirv-input-builtin.md

```diff
@@ -0,0 +1,159 @@
+# SPIR-V Input/Output built-ins
+
+ * Proposal: [NNNN](NNNN-spirv-input-builtin.md)
+ * Author(s): [Nathan Gauër](https://github.com/Keenuts)
+ * Status: **Design In Progress**
+
+## Introduction
+
+HLSL has semantic input/outputs parameters used to carry information
+to and from a shader, e.g: the system semantic `SV_GroupID` or `MY_SEMANTIC`,
+a user-defined semantic.
+
+In SPIR-V, those are translated to `Input` and `Output` variables,
+with either a `BuiltIn` or `Location` decoration.
+
+Example:
+
+```hlsl
+float4 vsmain(float4 Position : POSITION) : SV_POSITION
+{
+  return Position;
+}
+```
+
+Both `POSITION` and `SV_POSITION` will be represented as an `OpVariable`:
+
+- `POSITION` will have the `Input` storage class.
+- `SV_POSITION` will have the `Output` storage class.
+
+- `POSITION` will be decorated with `Location 0`.
+- `SV_POSITION` will be decorated with `BuiltIn Position`.
+
+In addition, users can define their own SPIR-V built-ins using inline SPIR-V:
+
+```hlsl
+[[vk::ext_builtin_input(/* NumWorkGroups */ 24)]]
+static const uint3 myBuiltIn;
+```
+
+This document explain how we plan on implementing those in Clang/LLVM.
+
+## Proposed Solution
+
+## Static behaving like an external & LLVM
+
+The current HLSL syntax for inline SPIR-V requires those built-ins to be
+`static [const]` variables. The rational was those variables were private to
+the invocation, and thus could be written/read without any concurrency issues.
+This was fine in DXC as we had our own AST->SPIR-V lowering.
+
+In Clang/LLVM, that's different: such static variable can be optimized in many
+ways:
+aggressive ways.
+ - A non-initialized `static const` is not allowed.
+ - Any dead-store to a `static int` can be removed.
+ - Meaning any pure computation parent to this store can be removed.
+
+Solving this can be done in 3 ways:
+ - Change the spec, allowing the syntax to be less "magical": use `extern`.
+ - Fixup the storage class of those variables in Clang.
+ - Emit an external constructor & destructor for those preventing optimization.
+
+I'd be in favor of the first solution, which correctly expresses in HLSL what
+those built-in variables are, see this
+[spec issue](https://github.com/microsoft/hlsl-specs/issues/350).
+If we emit those as `extern [thread_local]`, everything works out of the box,
+and HLSL feels less magical.
```

Comment:
That's where the `thread_local` comes in. Yes, a simple global `extern` is assumed to be shared across threads, but if we have an `extern thread_local`, then it would carry the semantic we want in HLSL, while being also find from LLVM's point of view.

Reply:
> We need to distinguish between how the variable works in llvm, and how it works in HLSL.

I'm thinking about this from the perspective of the person who is writing their code in HLSL. There is no `thread_local` in HLSL. The `static` keyword is used to indicate that a global variable is thread local. So I don't think the spec should be changed.

How it is represented in LLVM is an important issue, and I don't know what is best. I'll let @llvm-beanz  comment on what is more invasive in clang: static that is initialized or used in an intrinsic in the wrapper, or changing variable to extern thread local internally. This is in some ways orthogonal to changing the spec.